### PR TITLE
[RFC] Doc-check utility.

### PR DESCRIPTION
**Pull request by tjingrant**
cc @doru1004  @chentong319  @AlexandreEichenberger @tungld 

I am trying to gather comments about this utility called DocCheck.

It essentially provides utilities to ensure that documentation is up-to-date with the status of our software development. Consider the following use case:

For instance, MLIR and ONNF installation scripts are stored under utils directory, and they are used in our build bot to build MLIR/ONNF for testing, so we know for sure they work. Now we want to ensure that our main README.md contains build instructions that are (almost) the same copy of these two scripts used in our build bot. 

What I propose we do is to enforce that the build instructions included in README.md are always the same with the script under utils/install-mlir|onnf.sh using a directive `same-as-file` in README.md like below:

````
[same-as-file]: <> (utils/install-mlir.sh)
``` bash
git clone https://github.com/llvm/llvm-project.git
mkdir llvm-project/build
cd llvm-project/build
cmake -G Ninja ../llvm \
   -DLLVM_ENABLE_PROJECTS=mlir \
   -DLLVM_BUILD_EXAMPLES=ON \
   -DLLVM_TARGETS_TO_BUILD="host" \
   -DCMAKE_BUILD_TYPE=Release \
   -DLLVM_ENABLE_ASSERTIONS=ON \
   -DLLVM_ENABLE_RTTI=ON
cmake --build . --target check-mlir -- ${MAKEFLAGS}
```
````

A tool doc_check/check.py automatically checks whether the code section following this same-as-file directive contains code that is the same as the reference file. Such checking will be an integral stage of our continuous integration system.

I finished a preliminary implementation today and want to know whether you agree this is a good idea before moving on to making it more complete.

### CODE REVIEW

**Comment by tungld** [comment]
I like the idea. Could you please do the same thing for the session `Using ONNF` of `README.md` by checking the output of the command `onnf --help`?

**Comment by tjingrant** [comment]
@tungld yes that's exactly what I hope to do next! Implement another directive `same-as-stdout` which works similarly like `same-as-file`. Will follow up with another patch after this one lands.

File: doc/doc_check/check.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# ===-------------------- gen.py - Generate Documentation -----------------===//
```
Comment (doru1004) [inline doc/doc_check/check.py]: check.py

**Comment by chentong319** [comment]
To my understanding, this work is intended to "check" the consistency among .md files with our own python script. Can we generate those .md files with script to ensure consistency. I recalled that Samuel wrote a shell script, called universal_install, for openmp project to handle all the components. I guess the script can share some options across the components for consistency.
Will Doc-check target the consistency on options used in different component?

**Comment by tjingrant** [comment]
@chentong319 this work checks that the code included in documentation files is consistent with the code used in (unit) tests. One thing I'd like to clarify is that DocCheck doesn't check the consistency between .md files themselves, it checks the consistency between .md files and unit tests to make sure code snippets presented in documentation is always up-to-date.

Yes we can generate documentation with a script to achieve similar effect, but I feel like that's too intrusive to the workflow of documentation writing. One of the important aspect of Markdown is what-you-see-is-what-you-get, generating documentation defeats this desirable property of Markdown. 

>  for openmp project to handle all the components

Not sure what do you mean by handle all the components? Do you mean generating documentation for all components? Or do you mean building all components?

> Will Doc-check target the consistency on options used in different component?

Not sure how this comment relates to this patch. When you say "options", do you mean "build options"? I feel like we may be talking about different things. DocCheck does not intend to enforce the same set of options/code/string being used everywhere, it simply checks the code you put in places where you can't test directly (e.g., documentation) is the same as the code you used in a unit test. This guarantees that code snippets in the documentation are always tested and up-to-date with the codebase itself.

**Comment by chentong319** [comment]
@tjingrant  As you said, it only check whether the document and the unit test are the same.  I thought it is used to coordinate the build ONNF and MLIR,  which is not the goal of this change.

**Comment by doru1004** [comment]
I think this is a good idea to make it easy to catch things that are out of sync. Thanks @tjingrant for the PR! Please let me know when you're done adding all the features.

**Comment by tjingrant** [comment]
@doru1004 done. Please help review.
I've also tried using doc-check to check the documentation for doc-check; seems to work well.

**Comment by doru1004** [comment]
LGTM!

**Pull request merged**